Show notification if newer version has been released

anttimaki opened the pull request:
Release information is read from GitHub repo. To avoid slowing the
usage of the CLI too much, the request timeouts after three seconds.

The version number is read from the JSON response with regex to avoid
JSON deserialization. This requires that the release versions follow
the X.Y.Z semantic versioning.

In case of errors while reading the version of the latest release, the
method fails silently, since it's unlikely that the end user would be
interested in any error messages regarding these issues.

The version number of currently installed app is read from .csproj
file, meaning that the version number must be manually updated in the
code whenever a new version is released.

MythicManiac reviewed (approved):
Code-wise looks good enough to me, though I'm wondering how did you test this given that we have no releases yet? Probably used another repository as a test?

anttimaki commented:
Yes, I used another repository for testing purposes.

I'm currently looking for solutions like https://github.com/adamralph/minver , which may cause changes to versioning scheme, so I'll leave this PR open for now.

anttimaki commented:
@MythicManiac 

* Added a [new commit](https://github.com/thunderstore-io/thunderstore-cli/pull/24/commits/91982f3565b086e4421e9b52187b8de1f5f32146) which adds MinVer library for managing versioning via tags, so we don't need to update the project config file manually
* Rewrote [GetCurrentVersion](https://github.com/thunderstore-io/thunderstore-cli/pull/24/commits/2360844f4c008986d375696a4b7c054aae5084e2#diff-1d63bd23e24df93ef887ec1a8a8d8d6907959b73e0da44556b4cda5133adaaf3R12-R39) and partially the commit message, the second commit is otherwise untouched
* As a test, added tag 0.1.0-alpha.1

MythicManiac commented on Utils/MiscUtils.cs:
```diff
@@ -0,0 +1,73 @@
+        /// <exception cref="ArgumentException">Throw if version number not found</exception>
+        public static int[] ParseLatestVersion(string releaseJsonData)
+        {
+            var regex = new Regex(@"""tag_name"":""(\d+.\d+.\d+)""");
```
Do we need full semver support here now, or is the suffix (e.g. `-alpha`) not included in the binary?

anttimaki replied:
`GetCurrentVersion` reads the value from binary. This method reads the value from JSON metadata returned by GitHub API. My idea was that there would be no notifications about alpha-releases, i.e. this would throw the `ArgumentException` and the caller method would fail silently.

But now that I think about this, this would also fail silently is situations where user has e.g. `0.1.0`, there's a stable `0.2.0` release, but the latest one is `0.3.0-alpha.1`.

So I guess that if we want to support alpha-releases (instead of just tags), this needs to be switched so that we fetch information about more than just the latest release. It would also be required if there's going to be e,g, more than one supported major release at a time (e.g. releasing `1.2.3` and `2.0.2` that contain the same hotfix at the same time).

Would it be a good idea to support those cases at this time?

MythicManiac replied:
I'd say if it's not too much work (1d tops), then sure, would hate the update notification breaking entirely because someone pushes a suffixed version. But it's also not an irrecoverable state, as you can just push a new release with a proper tag format.

https://github.com/maxhauser/semver this is the closest parser I found with a quick search, but it also seems rather unmaintained. Still could be good for pointers at least?

anttimaki replied:
Well I think we just need to query a different endpoint, and parse release tags in a loop, and then pick the latest stable release, so it shouldn't take that long. And I'll see if I can do it without external dependencies first.

I still assume we don't want to show the notification for alpha-releases.

MythicManiac replied:
Sounds good to me, and yeah makes sense to only show full

anttimaki replied:
* Added a [comparer for semvers](https://github.com/thunderstore-io/thunderstore-cli/pull/24/commits/44b8d9d3809560723e8e21e0d822cb30e385485e#diff-a240d47b5472ff5ff9ca38a4665ec9cd04254a6dd79d4b30bdf1e13ceafaf6cbR1)
* Rewrote [ParseLatestVersion](https://github.com/thunderstore-io/thunderstore-cli/pull/24/commits/44b8d9d3809560723e8e21e0d822cb30e385485e#diff-1d63bd23e24df93ef887ec1a8a8d8d6907959b73e0da44556b4cda5133adaaf3R43) to process a string that contains multiple releases
* Renamed the [fetcher method](https://github.com/thunderstore-io/thunderstore-cli/pull/24/commits/44b8d9d3809560723e8e21e0d822cb30e385485e#diff-1d63bd23e24df93ef887ec1a8a8d8d6907959b73e0da44556b4cda5133adaaf3R63) and updated the endpoint URL, so now all releases are fetched, not just the latest. Also updated the place where the method is called.

Other places weren't touched.

MythicManiac replied:
Thanks 👍

MythicManiac reviewed (approved):
Looks good to me, nice work 👍 

As a side note, the SemVer comparer is the kind of thing that makes me wish we had a test framework set up on this project already. It _might_ be worth writing a test case for the SemVer comparer already, I feel like that would be doable in an hour or so. Up to you.

anttimaki commented:
I'll look into testing, but that'll be a separate PR.
